File: apple/Transponder/Transponder/LocationSyncService.swift
import Foundation
import CoreLocation

/// Service for syncing locations with the Transponder server
class LocationSyncService {
    private let identityStore: IdentityStore
    private let urlSession: URLSession

    init(identityStore: IdentityStore = .shared) {
        self.identityStore = identityStore
        self.urlSession = URLSession.shared
    }

    // MARK: - Result Types

    enum UploadResult {
        case success
        case error(String)
    }

    enum FetchResult {
        case success([FetchedLocation])
        case error(String)
    }

    enum ServerValidationResult {
        case valid(name: String, version: String)
        case invalidServer
        case networkError(String)
    }

    // MARK: - Server Validation

    func validateServer(url: String) async -> ServerValidationResult {
        let baseUrl = url.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard let requestUrl = URL(string: "\(baseUrl)/api/version") else {
            return .invalidServer
        }

        do {
            let (data, response) = try await urlSession.data(from: requestUrl)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return .invalidServer
            }

            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let name = json["name"] as? String,
                  let version = json["version"] as? String,
                  name == "coords" || name == "transponder" else {
                return .invalidServer
            }

            return .valid(name: name, version: version)
        } catch {
            return .networkError(error.localizedDescription)
        }
    }

    // MARK: - Add Friend with Location Upload

    enum AddFriendResult {
        case success                      // Friend added, no upload attempted
        case successWithUpload            // Friend added, upload succeeded
        case successUploadFailed(String)  // Friend added, but upload failed
        case addFriendFailed(String)      // Failed to add friend
    }

    /// Adds a friend and optionally uploads current location if auto-share is enabled
    func addFriendAndUploadLocation(
        pubkey: String,
        server: String,
        name: String,
        location: CLLocation?
    ) async -> AddFriendResult {
        // First, add the friend
        do {
            try addFriend(
                pubkey: pubkey,
                server: server,
                name: name,
                shareWith: true,
                fetchFrom: true
            )
        } catch {
            return .addFriendFailed(error.localizedDescription)
        }

        // Upload location if auto-share is enabled and we have location data
        if identityStore.autoShareEnabled, let location = location {
            let uploadResult = await uploadLocation(
                latitude: location.coordinate.latitude,
                longitude: location.coordinate.longitude,
                accuracy: Float(location.horizontalAccuracy),
                timestamp: UInt64(location.timestamp.timeIntervalSince1970 * 1000)
            )
            switch uploadResult {
            case .success:
                return .successWithUpload
            case .error(let message):
                return .successUploadFailed(message)
            }
        }

        return .success
    }

    // MARK: - Location Upload

    func uploadLocation(latitude: Double, longitude: Double, accuracy: Float, timestamp: UInt64) async -> UploadResult {
        guard let identity = identityStore.getIdentity() else {
            return .error("No identity configured")
        }

        let serverUrl = identityStore.serverUrl

        let location = Location(
            latitude: latitude,
            longitude: longitude,
            accuracy: accuracy,
            timestamp: timestamp
        )

        // Get friends we share location with
        let shareRecipients = getShareRecipients()

        do {
            let request = try prepareLocationUpload(
                identity: identity,
                location: location,
                friends: shareRecipients,
                myServer: serverUrl
            )

            let result = await execute(request: request)

            switch result {
            case .success:
                markUploadSuccess(timestamp: timestamp)
                return .success
            case .httpError(let code, let message):
                return .error("Server error: \(code) \(message)")
            case .networkError(let message):
                return .error("Network error: \(message)")
            }
        } catch CoreError.StaleLocation {
            // Location is older than last upload, skip silently
            print("LocationSync: Skipping stale location upload")
            return .success
        } catch {
            return .error("Upload failed: \(error.localizedDescription)")
        }
    }

    // MARK: - Location Fetch

    func fetchFriendLocations(friends: [Friend]) async -> FetchResult {
        guard let identity = identityStore.getIdentity() else {
            return .error("No identity configured")
        }

        if friends.isEmpty {
            return .success([])
        }

        do {
            let requests = prepareLocationFetch(friends: friends)
            var allLocations: [FetchedLocation] = []

            for request in requests {
                let result = await execute(request: request)
                if case .success(let data) = result {
                    let locations = try processFetchResponse(identity: identity, responseBody: data)
                    allLocations.append(contentsOf: locations)
                }
            }

            return .success(allLocations)
        } catch {
            return .error("Fetch failed: \(error.localizedDescription)")
        }
    }

    func fetchSelfLocation() async -> FetchResult {
        guard let identity = identityStore.getIdentity() else {
            return .error("No identity configured")
        }

        let serverUrl = identityStore.serverUrl
        let request = prepareSelfLocationFetch(identity: identity, server: serverUrl)

        let result = await execute(request: request)

        switch result {
        case .success(let data):
            do {
                let locations = try processFetchResponse(identity: identity, responseBody: data)
                return .success(locations)
            } catch {
                return .error("Failed to process response: \(error.localizedDescription)")
            }
        case .httpError(let code, let message):
            return .error("Server error: \(code) \(message)")
        case .networkError(let message):
            return .error("Network error: \(message)")
        }
    }

    /// Fetch locations for all friends we're tracking (fetch_from == true)
    /// Updates the cached locations in storage
    func fetchTrackedFriends() async -> FetchResult {
        guard let identity = identityStore.getIdentity() else {
            return .error("No identity configured")
        }

        let targets = getFetchTargets()
        if targets.isEmpty {
            return .success([])
        }

        do {
            let requests = prepareLocationFetch(friends: targets)
            var allLocations: [FetchedLocation] = []
            let now = UInt64(Date().timeIntervalSince1970 * 1000)

            for request in requests {
                let result = await execute(request: request)
                if case .success(let data) = result {
                    let locations = try processFetchResponse(identity: identity, responseBody: data)

                    // Update cached locations in storage
                    for loc in locations {
                        try? updateFriendLocation(
                            pubkey: loc.pubkey,
                            location: loc.location,
                            fetchedAt: now
                        )
                    }

                    allLocations.append(contentsOf: locations)
                }
            }

            return .success(allLocations)
        } catch {
            return .error("Fetch failed: \(error.localizedDescription)")
        }
    }

    // MARK: - HTTP Execution

    private enum HttpResult {
        case success(Data)
        case httpError(Int, String)
        case networkError(String)
    }

    private func execute(request: PreparedRequest) async -> HttpResult {
        guard let url = URL(string: request.url) else {
            return .networkError("Invalid URL")
        }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = request.method

        for (key, value) in request.headers {
            urlRequest.setValue(value, forHTTPHeaderField: key)
        }

        if !request.body.isEmpty {
            urlRequest.httpBody = request.body
        }

        do {
            let (data, response) = try await urlSession.data(for: urlRequest)

            guard let httpResponse = response as? HTTPURLResponse else {
                return .networkError("Invalid response")
            }

            if httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 {
                return .success(data)
            } else {
                let message = HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode)
                return .httpError(httpResponse.statusCode, message)
            }
        } catch {
            return .networkError(error.localizedDescription)
        }
    }
}

File: android/app/src/main/java/sh/bentley/transponder/LocationSyncService.kt
package sh.bentley.transponder

import kotlinx.coroutines.Deferred
import uniffi.transponder_core.CoreException
import uniffi.transponder_core.FetchedLocation
import uniffi.transponder_core.Friend
import uniffi.transponder_core.Location
import uniffi.transponder_core.addFriend
import uniffi.transponder_core.getFetchTargets
import uniffi.transponder_core.getShareRecipients
import uniffi.transponder_core.markUploadSuccess
import uniffi.transponder_core.prepareLocationFetch
import uniffi.transponder_core.prepareLocationUpload
import uniffi.transponder_core.prepareSelfLocationFetch
import uniffi.transponder_core.processFetchResponse
import uniffi.transponder_core.updateFriendLocation

/**
 * Service for syncing locations with the Transponder server.
 * Coordinates between Rust core (crypto) and HTTP layer (network).
 */
class LocationSyncService(
    private val identityStore: IdentityStore,
    private val http: TransponderHttp = TransponderHttp()
) {
    sealed class UploadResult {
        data object Success : UploadResult()
        data class Error(val message: String) : UploadResult()
    }

    sealed class FetchResult {
        data class Success(val locations: List<FetchedLocation>) : FetchResult()
        data class Error(val message: String) : FetchResult()
    }

    sealed class AddFriendResult {
        data object Success : AddFriendResult()              // Friend added, no upload attempted
        data object SuccessWithUpload : AddFriendResult()    // Friend added, upload succeeded
        data class SuccessUploadFailed(val message: String) : AddFriendResult()  // Friend added, upload failed
        data class AddFriendFailed(val message: String) : AddFriendResult()      // Failed to add friend
    }

    /**
     * Adds a friend and optionally uploads current location if auto-share is enabled.
     * Location is passed as a Deferred so it can resolve in the background while
     * the friend is being added.
     */
    suspend fun addFriendAndUploadLocation(
        pubkey: String,
        server: String,
        name: String,
        locationDeferred: Deferred<android.location.Location?>?
    ): AddFriendResult {
        // First, add the friend immediately (no waiting for location)
        try {
            addFriend(
                pubkey = pubkey,
                server = server,
                name = name,
                shareWith = true,
                fetchFrom = true
            )
        } catch (e: Exception) {
            return AddFriendResult.AddFriendFailed(e.message ?: "Unknown error")
        }

        // Now await the location (which has been resolving in background) and upload
        if (identityStore.autoShareEnabled && locationDeferred != null) {
            val location = locationDeferred.await()
            if (location != null) {
                val uploadResult = uploadLocation(
                    latitude = location.latitude,
                    longitude = location.longitude,
                    altitude = location.altitude,
                    accuracy = location.accuracy,
                    timestamp = location.time
                )
                return when (uploadResult) {
                    is UploadResult.Success -> AddFriendResult.SuccessWithUpload
                    is UploadResult.Error -> AddFriendResult.SuccessUploadFailed(uploadResult.message)
                }
            }
        }

        return AddFriendResult.Success
    }

    /**
     * Upload our current location to the server.
     */
    suspend fun uploadLocation(
        latitude: Double,
        longitude: Double,
        altitude: Double,
        accuracy: Float,
        timestamp: Long
    ): UploadResult {
        val identity = identityStore.getIdentity()
            ?: return UploadResult.Error("No identity configured")

        val uploadPubkeyHex = identity.ed25519Public.take(8).joinToString("") { "%02x".format(it.toByte()) }
        android.util.Log.d("LocationSync", "upload identity pubkey: $uploadPubkeyHex...")

        val serverUrl = identityStore.serverUrl
            ?: return UploadResult.Error("No server URL configured")

        // Get friends we should encrypt for (those with shareWith == true)
        val friends = getShareRecipients()

        val location = Location(
            latitude = latitude,
            longitude = longitude,
            altitude = altitude,
            accuracy = accuracy,
            timestamp = timestamp.toULong()
        )

        return try {
            val request = prepareLocationUpload(identity, location, friends, serverUrl)

            when (val result = http.execute(request)) {
                is TransponderHttp.Result.Success -> {
                    markUploadSuccess(timestamp.toULong())
                    UploadResult.Success
                }
                is TransponderHttp.Result.HttpError ->
                    UploadResult.Error("Server error: ${result.code} ${result.message}")
                is TransponderHttp.Result.NetworkError ->
                    UploadResult.Error("Network error: ${result.exception.message}")
            }
        } catch (e: CoreException.StaleLocation) {
            // Location is older than last upload, skip silently
            android.util.Log.d("LocationSync", "Skipping stale location upload")
            UploadResult.Success
        } catch (e: Exception) {
            UploadResult.Error("Upload failed: ${e.message}")
        }
    }

    /**
     * Fetch locations for all friends we're tracking (fetchFrom == true).
     * Updates the cached locations in storage.
     */
    suspend fun fetchTrackedFriends(): FetchResult {
        val identity = identityStore.getIdentity()
            ?: return FetchResult.Error("No identity configured")

        val targets = getFetchTargets()
        if (targets.isEmpty()) {
            return FetchResult.Success(emptyList())
        }

        return try {
            val requests = prepareLocationFetch(targets)
            val allLocations = mutableListOf<FetchedLocation>()
            val now = System.currentTimeMillis().toULong()

            for (request in requests) {
                when (val result = http.execute(request)) {
                    is TransponderHttp.Result.Success -> {
                        val locations = processFetchResponse(identity, result.body)

                        // Update cached locations in storage (only if we got a valid location)
                        for (loc in locations) {
                            // Only update if we successfully decrypted a location
                            // If decryption fails (e.g., friend removed us), keep showing last known location
                            if (loc.location != null) {
                                try {
                                    updateFriendLocation(loc.pubkey, loc.location, now)
                                } catch (e: Exception) {
                                    android.util.Log.e("LocationSync", "Failed to update location for ${loc.pubkey}: ${e.message}")
                                }
                            }
                        }

                        allLocations.addAll(locations)
                    }
                    is TransponderHttp.Result.HttpError -> {
                        android.util.Log.w("LocationSync", "HTTP error fetching friends: ${result.code}")
                    }
                    is TransponderHttp.Result.NetworkError -> {
                        android.util.Log.w("LocationSync", "Network error fetching friends: ${result.exception.message}")
                    }
                }
            }

            FetchResult.Success(allLocations)
        } catch (e: Exception) {
            FetchResult.Error("Fetch failed: ${e.message}")
        }
    }

    /**
     * Fetch locations for all friends.
     * Handles federation by grouping requests by server.
     */
    suspend fun fetchFriendLocations(friends: List<Friend>): FetchResult {
        val identity = identityStore.getIdentity()
            ?: return FetchResult.Error("No identity configured")

        if (friends.isEmpty()) {
            return FetchResult.Success(emptyList())
        }

        return try {
            // Prepare requests grouped by server
            val requests = prepareLocationFetch(friends)

            // Execute all requests and collect results
            val allLocations = mutableListOf<FetchedLocation>()

            for (request in requests) {
                when (val result = http.execute(request)) {
                    is TransponderHttp.Result.Success -> {
                        val locations = processFetchResponse(identity, result.body)
                        allLocations.addAll(locations)
                    }
                    is TransponderHttp.Result.HttpError -> {
                        // Log but continue with other servers
                    }
                    is TransponderHttp.Result.NetworkError -> {
                        // Log but continue with other servers
                    }
                }
            }

            FetchResult.Success(allLocations)
        } catch (e: Exception) {
            FetchResult.Error("Fetch failed: ${e.message}")
        }
    }

    /**
     * Fetch our own location from the server.
     * Used to verify what the server has stored for us.
     */
    suspend fun fetchSelfLocation(): FetchResult {
        val identity = identityStore.getIdentity()
            ?: return FetchResult.Error("No identity configured")

        val fetchPubkeyHex = identity.ed25519Public.take(8).joinToString("") { "%02x".format(it.toByte()) }
        android.util.Log.d("LocationSync", "fetch identity pubkey: $fetchPubkeyHex...")

        val serverUrl = identityStore.serverUrl
            ?: return FetchResult.Error("No server URL configured")

        return try {
            val request = prepareSelfLocationFetch(identity, serverUrl)

            when (val result = http.execute(request)) {
                is TransponderHttp.Result.Success -> {
                    android.util.Log.d("LocationSync", "fetchSelfLocation response size: ${result.body.size}")
                    android.util.Log.d("LocationSync", "fetchSelfLocation response: ${result.body.decodeToString()}")
                    val locations = processFetchResponse(identity, result.body)
                    android.util.Log.d("LocationSync", "fetchSelfLocation parsed ${locations.size} locations")
                    locations.forEach { loc ->
                        android.util.Log.d("LocationSync", "  pubkey=${loc.pubkey.take(8)}..., hasLocation=${loc.location != null}")
                    }
                    FetchResult.Success(locations)
                }
                is TransponderHttp.Result.HttpError ->
                    FetchResult.Error("Server error: ${result.code} ${result.message}")
                is TransponderHttp.Result.NetworkError ->
                    FetchResult.Error("Network error: ${result.exception.message}")
            }
        } catch (e: Exception) {
            FetchResult.Error("Fetch failed: ${e.message}")
        }
    }
}
